2.3-free-monoids.md #14

arioston opened the pull request:
Pre Work 2.3-free-monoids.md #14

arioston committed:
[Pre Work] 2.3-free-monoids.md #14

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,87 @@
+2 * 3 = 6
+[2] ++ [3] = [2, 3] // not the same as [6]
+```
+```kotlin:ank:silent
````
Change this to kotlin, as this is not intended to be compiled. 
I will add more explanations later to the README.

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,87 @@
+```Haskell
+2 * 3 = 6
+```
+```kotlin:ank:silent
````
change to kotlin

nomisRev replied:
```suggestion
```kotlin
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,87 @@
+```Haskell
+q = U h . p
+```
+```kotlin
````
U is a Functor in your snippet it could be a variable uh. You may change it to U(h).

i-walker reviewed (commented):
Don't forget your imports.

arioston committed:
correction of the imports

i-walker commented:
Feel free to add more commits and I will help you along the way until you're done with the chapter. 
With that, I mean to integrate it with Arrow's [Free](https://github.com/arrow-kt/arrow/blob/master/modules/free/arrow-free-data/src/main/kotlin/arrow/free/Free.kt).

We can also work on it together (I am new to Free, too) here are some of the resources I found so far.
[A nice'ish explaination](https://stackoverflow.com/questions/13352205/what-are-free-monads)
[In the Kotlin slack](https://kotlinlang.slack.com/archives/C5UPMM0A0/p1554022508033000)

raulraja commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,87 @@
+  mappend = (++)
+```
+```kotlin:ank:silent
+class ListMonoid<M> : Monoid<List<M>> {
````
Use `A` as a type arg instead of `M` since that is the Arrow convention for concrete not kinded type args

arioston commented:
@i-walker Thanks, I was trying to figure out how to incorporate Arrow's Free.  I will appreciate if we can work together.

i-walker commented:
@arioston If you have any questions whatsoever don't hesitate to ask in the #arrow-contributor slack group. 
> @i-walker Thanks, I was trying to figure out how to incorporate Arrow's Free. I will appreciate if we can work together.

arioston committed:
Correction on free monoid implementation.
I have used [FreeMonoid](https://github.com/typelevel/spire/blob/master/core/src/main/scala/spire/algebra/free/FreeMonoid.scala) as reference.

i-walker commented on src/main/ank/2.3-free-monoids.md:
```diff
@@ -4,9 +4,20 @@ class Monoid m where
-interface Monoid<M> {
-  fun mempty() : M
-  fun mappend(a1: M, a2: M): M
+package arrow.typeclasses
```
delete this line, it's unnecessary.
>  package arrow.typeclasses

nomisRev replied:
```suggestion
```

nomisRev replied:
```suggestion
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
```diff
@@ -4,9 +4,20 @@ class Monoid m where
-  fun mappend(a1: M, a2: M): M
+package arrow.typeclasses
+
+class ForMonoid private constructor() { companion object }
```
Even though, it's directly from the arrow doc's the reader won't be familiar with either of the following type classes or the reason, why we use MonoidOf. 
Just delete the lines 9 - 11 and 
remove Semigroup<A>, MonoidOf<A> and the combineAll functions.

i-walker commented on src/main/ank/2.3-free-monoids.md:
```diff
@@ -4,9 +4,20 @@ class Monoid m where
+  fun empty(): A
+
+  fun Collection<A>.combineAll(): A =
+    if (isEmpty()) empty() else reduce { a, b -> a.combine(b) }
```
Instead, write the combine function explicitly.

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -15,10 +26,35 @@ instance Monoid [a] where
-  override fun mempty() = listOf<M>()
-  override fun mappend(a1: List<M>, a2: List<M>) = a1 + a2
+```kotlin
+package free
````
remove package declaration

nomisRev replied:
```suggestion
```

nomisRev replied:
```suggestion
package free
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -28,7 +64,7 @@ class ListMonoid<M> : Monoid<List<M>> {
 ```
 ```kotlin
 2 * 3 = 6
-listOf(2) + listOf(3) == listOf(2, 3) // not the same as listOf(6)
````
keep it simple. revert the changes to 
> listOf(2) + listOf(3) == listOf(2, 3) // not the same as listOf(6)

nomisRev replied:
```suggestion
listOf(2) + listOf(3) == listOf(2, 3) // not the same as listOf(6)
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -44,7 +80,7 @@ h(a * b) == h(a) * h(b)
 [2] ++ [3] = [2, 3]
 ```
 ```kotlin
-listOf(2) + listOf(3) == listOf(2, 3)
````
here the same as above

nomisRev replied:
```suggestion
listOf(2) + listOf(3) == listOf(2, 3)
```

nomisRev commented on src/main/ank/2.3-free-monoids.md:
```diff
@@ -0,0 +1,123 @@
+import arrow.higherkind
+import arrow.typeclasses.Monoid
+
+operator fun <A>FreeMonoid<A>.plus(b: FreeMonoid<A>) = FreeMonoid.just(this.terms + b.terms)
```
```suggestion
operator fun <A> FreeMonoid<A>.plus(b: FreeMonoid<A>) = FreeMonoid.just(this.terms + b.terms)
```

nomisRev commented on src/main/ank/2.3-free-monoids.md:
```diff
@@ -0,0 +1,123 @@
+@higherkind
+sealed class FreeMonoid<S>(val terms: List<S>): FreeMonoidOf<S> {
+
+    internal data class Pure<A>(val a: List<A>) : FreeMonoid<A>(a)
```
Any reason for this to be an ADT? Can't we do `class FreeMonoid<S>(val terms: List<S>): FreeMonoidOf<S>` instead?

i-walker replied:
I suggest you @arioston check both Links to integrate the free monoid better with Arrow. 
- https://medium.com/@olxc/free-monads-explained-pt-1-a5c45fbdac30
- https://typelevel.org/cats/datatypes/freemonad.html

Feel free to ask questions.

arioston committed:
Correction after review:
 1. remove imports and package name
 2. remove ADT
 3. remove Semigroup and MonoidOf
 4. remove combineAll functions
 5. rollback of  FreeMonoid.just to listOf in examples

arioston committed:
Changing `kotlin:ank:silent` to `Kotlin`

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,107 @@
+  mappend = (++)
+```
+```kotlin
+import arrow.higherkind
````
This is ideally, what you're trying to do
`
@extension
interface Monoid<A>: MonoidOf<List<A>>{
  fun empty() = emptyList()
  fun List<A>.combine(other: List<A>) = this + other
}
`

i-walker replied:
These are the imports
```
import arrow.extension
import arrow.typeclasses.MonoidOf
```

arioston committed:
Monoid Implementation

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,90 @@
+  mempty = []
+  mappend = (++)
+```
+```kotlin
````
```suggestion
```kotlin:ank:playground
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
```diff
@@ -0,0 +1,90 @@
+import arrow.extension
+import arrow.typeclasses.MonoidOf
+
+@extension interface Monoid<A>: MonoidOf<List<A>> {
```
```suggestion
@extension 
interface Monoid<A>: MonoidOf<List<A>> {
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,90 @@
+p :: x -> U m
+```
+```kotlin
+val p : (X) -> Kind<U, M>
````
```
from the book:
We start with a set of generators, x. That's a set in Set. 
The pattern we are going to match consists of monoid m - an object of Mon - and a function p in Set,
where U is our forgetful functor from Mon to Set. 
This is a weird heterogeneous pattern - half in Mon and half in Set.
The idea is that the function p will identify the set of generators inside the X-ray image of m. 
(Page 212)
```
@nomisRev: Adjusted to his response
```suggestion
fun <A, U, M> p(x: A): Kind<U, M>
```

nomisRev replied:
This is actually not valid Kotlin code.

```
fun <X, U, M> p(a: A): Kind<U, M>
```

or

```
interface Example<X, U, M> {
   val p : (X) -> Kind<U, M>
}
```

There is no way to use generics unless they are defined in a function, or the lambda lives inside a class or interface that defines those generics.

i-walker committed:
Formatting Monoid

Co-Authored-By: arioston <ariostonj@hotmail.com>

i-walker committed:
Change kotlin to kotlin:ank:playground

Co-Authored-By: arioston <ariostonj@hotmail.com>

i-walker commented on src/main/ank/2.3-free-monoids.md:
```diff
@@ -0,0 +1,91 @@
+import arrow.typeclasses.MonoidOf
+
+@extension 
+interface Monoid<A>: MonoidOf<List<A>> {
```
```suggestion
interface Monoid<A>: MonoidOf<ForListK> {
```

pakoito replied:
That shouldn't be the case.

i-walker replied:
I thought ...Of<HKT> expects a HKT?
At least, that is what I've been said.

pakoito replied:
Monoid is aggregation of regular values, check the existing implementation of `ListK`.

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,91 @@
+  mappend = (++)
+```
+```kotlin:ank:playground
+import arrow.extension
````
```suggestion
import arrow.data.ForListK
import arrow.extension
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,91 @@
+q :: x -> U n
+```
+```kotlin
+val q: (X) -> Kind<U, N>
````
```suggestion
fun <A, U, N> q(x: A): Kind<U, N>
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,91 @@
+q = U(h) . p
+```
+```kotlin
+val q = uh compose p
````
```suggestion
val q = `Uh` compose p
```

i-walker commented on src/main/ank/2.3-free-monoids.md:
````diff
@@ -0,0 +1,91 @@
+h :: m -> n
+```
+```kotlin
+val h: (M) -> N
````
```suggestion
fun <M, N> h(m: M): N
```

i-walker committed:
Correction on the function definition

Co-Authored-By: arioston <ariostonj@hotmail.com>

i-walker committed:
Uppercase funtction

Co-Authored-By: arioston <ariostonj@hotmail.com>

i-walker committed:
Correction on the function definition

Co-Authored-By: arioston <ariostonj@hotmail.com>

i-walker committed:
Correction on the function definition

Co-Authored-By: arioston <ariostonj@hotmail.com>

i-walker commented:
😊😊🥳 Thank you good Job. 💪🏾
BTW you're one of my first "reviewees". I hope you liked it.
